sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3580,17 +3584,24 @@ pi_result piEventRetain(pi_event Event) {
 pi_result piEventRelease(pi_event Event) {
   assert(Event);
   if (--(Event->RefCount) == 0) {
+    // By default, cleanupAfterEvent will reuse the fence and the commandList
+    // associated with event. This is a good idea if the event will be alive.
+    // But, we are deallocating event in this function, so reusing the
+    // associated fence and commandList will cause memory leak.
+    // Note that we still need to release the kernel associated with the event.
+    // So, before calling cleanupAfterEvent, we have to indicate that the event
+    // is being removed by setting the Queue member to nullptr.
+    Event->Queue = nullptr;
```

Comment:
This is practically killing command-list caching and you will see performance regressions. Also I don't understand why not to reclaim the command-list and fence to the cache just before destroying the event. The cache is in the Queue and should get properly destroyed at the time the queue is about to be deleted.

Reply:
The problem is that the pi_event has a data member pi_queue.
This causes destruction of a pi_event to call a destruction of the pi_queue member, which in turn calls cleanupAfterEvent that will try to put the the fence and command-list in the cache.
This causes a memory leak because those objects in the cache are never be called with pi_release APIs.
In this function, we are destructing the pi_event obejct, so we don't need to cache the associated command-list and fence into the cache.

Reply:
> This is practically killing command-list caching
>> reusing command-list and fence is a good idea as long as the associated queue is still in use, but should avoid if the associated queue is already deallocated because it will cause a seg fault.

We should continue to reuse command-list/fence from the events being destroyed, and thus need to find other ways for fixing the problem you are seeing.

>> The problem is that a pi_event has an associated data member pi_queue.
But, if that pi_queue is already released, pi_event does not know if the associated pi_queue is already deallocated.

There was an earlier proposal to retain queue when event is linking to a queue, and release queue when event is deleted. So, effectively the queue is going to be deleted after all events in it are deleted.

Reply:
> There was an earlier proposal to retain queue when event is linking to a queue, and release queue when event is deleted. So, effectively the queue is going to be deleted after all events in it are deleted.

That won't solve the issue of pi_event being not notified of the pi_queue released. I already tried the "retain" approach in https://github.com/intel/llvm/pull/2905, and it did not solve this issue.

Reply:
> That won't solve the issue of pi_event being not notified of the pi_queue released. 

What do you mean by "notified" and for which events? I believe the proper way is to use retain/release to identify the internal usage of objects.

Reply:
I just meant that I tried retain/release approach in PR #2905, but it did not work.
Lots of LIT tests failed.
Next step:
1. Investigate why LIT tests failed in PR #2905. It looks like there are still the case piEventRelease occurs after piQueueRelease.
2. Investigate if I can keep the record of pi_queue without acutall "delete pi_queue", and deallocate them all at once in piTearDown. 

Please let me know if I am going to the wrong direction.

Reply:
I actually fixed the issue in #2905. 😃 
I will close this PR if you approve #2905.

Reply:
I like #2905, please close this one.